share/namespace.go

```diff
@@ -0,0 +1,134 @@
+package share
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+
+	appns "github.com/celestiaorg/celestia-app/pkg/namespace"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+// Various reserved namespaces.
+var (
+	MaxReservedNamespace     = appns.MaxReservedNamespace.Bytes()
+	ParitySharesNamespace    = appns.ParitySharesNamespace.Bytes()
+	TailPaddingNamespace     = appns.TailPaddingNamespace.Bytes()
+	ReservedPaddingNamespace = appns.ReservedPaddingNamespace.Bytes()
+	TxNamespace              = appns.TxNamespace.Bytes()
+	PayForBlobNamespace      = appns.PayForBlobNamespace.Bytes()
+)
+
+// Namespace represents namespace of a Share.
+// Consists of version byte and namespace ID.
+type Namespace []byte
+
+// NamespaceFromBytes converts bytes into Namespace and validates it.
+func NamespaceFromBytes(b []byte) (Namespace, error) {
+	n := Namespace(b)
+	return n, n.Validate()
+}
+
+// Version reports version of the Namespace.
+func (n Namespace) Version() byte {
+	return n[appns.NamespaceVersionSize-1]
+}
+
+// ID reports version of the Namespace.
+func (n Namespace) ID() namespace.ID {
+	return namespace.ID(n[appns.NamespaceVersionSize:])
+}
+
+// String stringifies the Namespace.
+func (n Namespace) String() string {
+	return hex.EncodeToString(n)
+}
+
+// Validate checks if the namespace is correct.
+func (n Namespace) Validate() error {
+	if len(n) != NamespaceSize {
+		return fmt.Errorf("invalid namespace length: %v must be %v", len(n), NamespaceSize)
+	}
+	if n.Version() != appns.NamespaceVersionZero && n.Version() != appns.NamespaceVersionMax {
+		return fmt.Errorf("unsupported namespace version %v", n.Version())
+	}
+	if len(n.ID()) != appns.NamespaceIDSize {
+		return fmt.Errorf("unsupported namespace id length: id %v must be %v bytes but it was %v bytes",
+			n.ID(), appns.NamespaceIDSize, len(n.ID()))
+	}
+	if n.Version() == appns.NamespaceVersionZero && !bytes.HasPrefix(n.ID(), appns.NamespaceVersionZeroPrefix) {
+		return fmt.Errorf("unsupported namespace id with version %v. GetNamespace %v must start with %v leading zeros",
+			n.Version(), n.ID(), len(appns.NamespaceVersionZeroPrefix))
+	}
+	return nil
+}
+
+// ValidateBlobNamespace returns an error if this namespace is not a valid blob namespace.
+func (n Namespace) ValidateBlobNamespace() error {
```

Comment:
also check for Tx Namespace and PFB namespace?

Reply:
Blob namespaces are the namespaces that can be specified by users when publishing a blob. The Tx namespace and PFB namespace are reserved namespaces so they can't be specified by users when publishing a blob.

If the tx namespace or PFB namespace is specified, it would be caught by https://github.com/celestiaorg/celestia-node/pull/2367/files/83977c48c94c4197dda01fdadefefc7e56660ed9#diff-46c85ee9c611a9b9c56eb104a6dca5bcbc6b3961683cdf0c941704b18496084dR71-R74

Reply:
Ah, right, as reserved means PFB and TX

Reply:
Ok, so I will do additional Validate like func that also restrict access for padding/erasured data

Reply:
Added